nixos/modules/services/networking/wireguard.nix

```diff
@@ -419,6 +483,116 @@ let
+      "~@memlock" "~@resources" "~@setuid" "~@timer"
+    ];
+    RestrictRealtime = true;
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    UMask = 0077;
+    ProtectHostname = true;
+    ProcSubset = "pid";
+  };
+
+  generatePeersAnnouncingSocket = name: values:
+    nameValuePair "wireguard-${name}-peers-announcing"
+      {
+        enable = values.peersAnnouncing.enable;
+        listenStreams = [(toString values.peersAnnouncing.port)];
+        socketConfig.Accept = true;
+        socketConfig.BindToDevice = name;
+        socketConfig.IPAddressAllow = map (v: v.allowedIPs) values.peers;
+        socketConfig.IPAddressDeny = "any";
+        socketConfig.MaxConnectionsPerSource = 1;
+        socketConfig.ReusePort = true;
+        wantedBy = [ "sockets.target" ];
+      };
+
+  generatePeersAnnouncingUnit = name: values:
+    nameValuePair "wireguard-${name}-peers-announcing@"
+      {
+        description = "WireGuard Peers Announcing - ${name}";
+        requires = [ "wireguard-${name}.service" ];
+        after = [ "wireguard-${name}.service" ];
+        path = with pkgs; [ wireguard-tools ];
+
+        serviceConfig = mkMerge [
+          peerUpdateSecurity
+          {
+            Type = "simple";
+            StandardInput = "null";
+            StandardOutput = "socket";
+            RestrictAddressFamilies = "";
+          }
+          (mkIf (values.interfaceNamespace != null)
+            { NetworkNamespacePath = "/var/run/netns/${values.interfaceNamespace}"; })
+        ];
+
+        # Send each configured peer, keeping only the fields: public_key and endpoint.
+        script = "wg show '${name}' dump | tail -n +2 | cut -f1,3";
+      };
+
+  generateEndpointReceivingUnit = { interfaceName, interfaceCfg, peer }: let
+      dynamicRefreshEnabled = peer.dynamicEndpointRefreshSeconds != 0;
+      peerService = peerUnitServiceName interfaceName peer.publicKey dynamicRefreshEnabled;
+    in
+    nameValuePair "wireguard-${interfaceName}-endpoint-receiving-${keyToUnitName peer.publicKey}"
+      {
+        description = "WireGuard ${interfaceName} Endpoint Receiving from ${peer.publicKey}";
+        requires = [ "${peerService}.service" ];
+        after = [ "${peerService}.service" ];
+        wantedBy = [ "${peerService}.service" ];
+        path = with pkgs; [ netcat-gnu wireguard-tools ];
+
+        serviceConfig = mkMerge [
+          peerUpdateSecurity
+          {
+            Type = "simple";
+            IPAddressAllow = [ peer.endpointReceiving.addr ];
+            RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_NETLINK" ];
+            Restart = "on-failure";
+          }
+          (mkIf (interfaceCfg.interfaceNamespace != null)
+            { NetworkNamespacePath = "/var/run/netns/${interfaceCfg.interfaceNamespace}"; })
+        ];
+
+        # Query the peer announcing other peers,
+        # for setting the current endpoint of all configured peers.
+        script = ''
```

Comment:
This is some nontrivial level of bash and quoting.

Given that Wireguard is highly security relevant, would you be up for doing the same in a safer-by-default script like Python instead?

Reply:
Sorry, I am no Pythoner, I personnaly would not be able to produce something more secure than this Bash script, in which I see no obvious problem.

AFAIU, because the Bash script is initiating the connection, inside the WireGuard tunnel, and only to the first hop, the input validation is not needed from a security point of view (unless the announcing peer is itself compromised). If that's correct, then that validation is here only from a feature point of view (updating only the peers configured in the peers' NixOS and not all those known by the announcing peer).
But feel free to point out if I'm mistaken or forgetting something here.

I have added timeouting (`-t`) and input limiting (`-n`) flags to the `read` in order to be more resilient/sane, but this is likely useless.